File: go/tools/file_io.go
package tools

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

func makeReadFileTool(sdPtr *string, readOnlyPaths []string) Tool {
	return Tool{
		Name:        "read_file",
		Description: "Read lines from a file (max 400 lines at a time, use offset to paginate)",
		Permission:  PermissionReadOnly,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the file to read",
				},
				"offset": map[string]any{
					"type":        "integer",
					"description": "Line number to start from (1-indexed, default 0 = beginning)",
				},
				"max_lines": map[string]any{
					"type":        "integer",
					"description": "Maximum lines to read starting from offset (default 200)",
				},
			},
			"required": []string{"path"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			resolved, err := ResolvePath(raw, *sdPtr, readOnlyPaths...)
			if err != nil {
				return "", err
			}

			offset := 0
			if v := getIntArg(args, "offset"); v >= 0 {
				offset = int(v)
				if offset < 0 {
					offset = 0
				}
			}
			maxLines := 200
			if v := getIntArg(args, "max_lines"); v > 0 {
				maxLines = int(v)
				if maxLines < 1 {
					maxLines = 1
				}
			}

			file, err := os.Open(resolved)
			if err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to open file: %s", resolved)}
			}
			defer file.Close()

			var result strings.Builder
			scanner := bufio.NewScanner(file)
			lineNum := 0
			count := 0

			// Skip lines before offset
			for lineNum < offset && scanner.Scan() {
				lineNum++
			}

			for scanner.Scan() && count < maxLines {
				lineNum++
				result.WriteString(scanner.Text())
				result.WriteString("\n")
				count++
			}

			// Check for more lines
			hasMore := false
			if scanner.Scan() {
				hasMore = true
			}

			if hasMore {
				result.WriteString(fmt.Sprintf("...(truncated, >%d lines from offset %d)", maxLines, offset))
			} else if count == 0 && offset > 0 {
				result.WriteString(fmt.Sprintf("(offset %d is beyond end of file)", offset))
			}

			return result.String(), nil
		},
	}
}

func makeReadFileLinesTool(sdPtr *string, readOnlyPaths []string) Tool {
	return Tool{
		Name:        "read_file_lines",
		Description: "Read specific line ranges from a file. Returns lines prefixed with line numbers. Use this when you know the line numbers you want (e.g. after a grep match at line 52, read lines 45-78). For reading from an offset, use read_file instead.",
		Permission:  PermissionReadOnly,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Path to the file",
				},
				"start_line": map[string]any{
					"type":        "integer",
					"description": "First line to read (1-indexed, default 1)",
				},
				"end_line": map[string]any{
					"type":        "integer",
					"description": "Last line to read (inclusive). If omitted, reads to end of file (capped by max_lines).",
				},
				"max_lines": map[string]any{
					"type":        "integer",
					"description": "Maximum lines to return (default 200, max 500)",
				},
			},
			"required": []string{"path"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			resolved, err := ResolvePath(raw, *sdPtr, readOnlyPaths...)
			if err != nil {
				return "", err
			}

			startLine := 1
			if v := getIntArg(args, "start_line"); v > 0 {
				startLine = int(v)
				if startLine < 1 {
					startLine = 1
				}
			}
			endLine := 0
			if v := getIntArg(args, "end_line"); v >= 0 {
				endLine = int(v)
			}
			maxLines := 200
			if v := getIntArg(args, "max_lines"); v > 0 {
				maxLines = int(v)
				if maxLines < 1 {
					maxLines = 1
				}
				if maxLines > 500 {
					maxLines = 500
				}
			}

			if endLine != 0 && endLine < startLine {
				return "", &ToolError{Message: "end_line must be >= start_line"}
			}

			file, err := os.Open(resolved)
			if err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to open file: %s", resolved)}
			}
			defer file.Close()

			var result strings.Builder
			scanner := bufio.NewScanner(file)
			lineNum := 0
			count := 0

			// Skip lines before startLine
			for lineNum < startLine-1 && scanner.Scan() {
				lineNum++
			}

			maxToRead := maxLines
			if endLine != 0 {
				range_ := endLine - startLine + 1
				if range_ < maxToRead {
					maxToRead = range_
				}
			}

			for count < maxToRead && scanner.Scan() {
				lineNum++
				result.WriteString(fmt.Sprintf("%d: %s\n", lineNum, scanner.Text()))
				count++
			}

			// Detect if there are more lines
			hasMore := false
			remaining := 0
			if endLine != 0 && lineNum < endLine {
				hasMore = true
				for scanner.Scan() {
					remaining++
				}
			} else if scanner.Scan() {
				hasMore = true
				remaining = 1
				for scanner.Scan() {
					remaining++
				}
			}

			if hasMore {
				result.WriteString(fmt.Sprintf("...(truncated, >%d lines from line %d)", count+remaining, startLine))
			} else if count == 0 && startLine > 1 {
				result.WriteString(fmt.Sprintf("(start_line %d is beyond end of file)", startLine))
			}

			return result.String(), nil
		},
	}
}

func makeWriteFileTool(sdPtr *string) Tool {
	return Tool{
		Name:        "write_file",
		Description: "Write content to a file, creating parent directories if needed",
		Permission:  PermissionWrite,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "File path",
				},
				"content": map[string]any{
					"type":        "string",
					"description": "Content to write",
				},
			},
			"required": []string{"path", "content"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			content, _ := args["content"].(string)

			resolved, err := ResolvePath(raw, *sdPtr)
			if err != nil {
				return "", err
			}

			// Create parent directories
			parent := filepath.Dir(resolved)
			if err := os.MkdirAll(parent, 0755); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to create parent directories: %s", err)}
			}

			if err := os.WriteFile(resolved, []byte(content), 0644); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to write file: %s", err)}
			}

			return fmt.Sprintf("ok (%d bytes written)", len(content)), nil
		},
	}
}

func makeEditFileTool(sdPtr *string) Tool {
	return Tool{
		Name:        "edit_file",
		Description: "Edit a file by searching for an exact string and replacing it. The search string must match exactly once in the file — this ensures edits are safe and unambiguous. Use this to make targeted surgical edits instead of rewriting entire files with write_file.",
		Permission:  PermissionWrite,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "File path to edit",
				},
				"search": map[string]any{
					"type":        "string",
					"description": "Exact string to search for; must match exactly once in the file. Include surrounding context (unique nearby lines) to guarantee a single match.",
				},
				"replace": map[string]any{
					"type":        "string",
					"description": "String to replace the matched occurrence with",
				},
			},
			"required": []string{"path", "search", "replace"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			search, _ := args["search"].(string)
			replace, _ := args["replace"].(string)

			if search == "" {
				return "", &ToolError{Message: "search string is required"}
			}

			resolved, err := ResolvePath(raw, *sdPtr)
			if err != nil {
				return "", err
			}

			// Read file
			data, err := os.ReadFile(resolved)
			if err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to read file: %s", resolved)}
			}
			content := string(data)

			// Count occurrences
			count := strings.Count(content, search)
			if count == 0 {
				return "", &ToolError{Message: "Search string not found in file (0 matches). Use read_file or grep_files to verify the file contents."}
			}
			if count > 1 {
				return "", &ToolError{Message: fmt.Sprintf("Search string found %d times in file (expected exactly 1). Include more surrounding context in the search string to uniquely identify the location.", count)}
			}

			// Find position for line number computation
			pos := strings.Index(content, search)
			if pos < 0 {
				return "", &ToolError{Message: "unexpected error: search string not found after count check"}
			}

			// Replace
			newContent := strings.Replace(content, search, replace, 1)

			// Write back
			if err := os.WriteFile(resolved, []byte(newContent), 0644); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to write file: %s", err)}
			}

			// Compute line number
			lineNum := 1
			for i := 0; i < pos; i++ {
				if content[i] == '\n' {
					lineNum++
				}
			}

			return fmt.Sprintf("ok (replaced 1 occurrence at line %d, %d bytes -> %d bytes)", lineNum, len(search), len(replace)), nil
		},
	}
}

func makeDeleteFileTool(sdPtr *string) Tool {
	return Tool{
		Name:        "delete_file",
		Description: "Delete a file",
		Permission:  PermissionWrite,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "File path to delete",
				},
			},
			"required": []string{"path"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			resolved, err := ResolvePath(raw, *sdPtr)
			if err != nil {
				return "", err
			}

			info, err := os.Stat(resolved)
			if err != nil {
				if os.IsNotExist(err) {
					return "", &ToolError{Message: fmt.Sprintf("File not found: %s", resolved)}
				}
				return "", &ToolError{Message: fmt.Sprintf("Cannot access file: %s", resolved)}
			}
			if info.IsDir() {
				return "", &ToolError{Message: fmt.Sprintf("Not a regular file: %s", resolved)}
			}

			size := info.Size()
			if err := os.Remove(resolved); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to delete file: %s", err)}
			}

			return fmt.Sprintf("ok (deleted %s, %d bytes)", resolved, size), nil
		},
	}
}

func makeMoveFileTool(sdPtr *string) Tool {
	return Tool{
		Name:        "move_file",
		Description: "Move or rename a file from source to destination. Works for both same-directory renames and cross-directory moves. Will not overwrite an existing destination.",
		Permission:  PermissionWrite,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"source": map[string]any{
					"type":        "string",
					"description": "Source file path",
				},
				"destination": map[string]any{
					"type":        "string",
					"description": "Destination file path",
				},
			},
			"required": []string{"source", "destination"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			srcRaw, _ := args["source"].(string)
			dstRaw, _ := args["destination"].(string)

			if srcRaw == "" || dstRaw == "" {
				return "", &ToolError{Message: "source and destination are required"}
			}

			src, err := ResolvePath(srcRaw, *sdPtr)
			if err != nil {
				return "", err
			}
			dst, err := ResolvePath(dstRaw, *sdPtr)
			if err != nil {
				return "", err
			}

			if _, err := os.Stat(src); os.IsNotExist(err) {
				return "", &ToolError{Message: fmt.Sprintf("Source not found: %s", src)}
			}
			if _, err := os.Stat(dst); err == nil {
				return "", &ToolError{Message: fmt.Sprintf("Destination already exists: %s", dst)}
			}

			// Create parent directories of destination
			parent := filepath.Dir(dst)
			if err := os.MkdirAll(parent, 0755); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to create parent directories: %s", err)}
			}

			// Rename (may fail cross-device)
			if err := os.Rename(src, dst); err != nil {
				// Cross-device link: fall back to copy + delete
				if strings.Contains(err.Error(), "invalid cross-device link") || strings.Contains(err.Error(), "cross-device") {
					if err := copyFile(src, dst); err != nil {
						return "", &ToolError{Message: fmt.Sprintf("Failed to copy across devices: %s", err)}
					}
					if err := os.Remove(src); err != nil {
						os.Remove(dst) // clean up destination
						return "", &ToolError{Message: fmt.Sprintf("Failed to remove source after copy: %s", err)}
					}
				} else {
					return "", &ToolError{Message: fmt.Sprintf("Failed to move file: %s", err)}
				}
			}

			return fmt.Sprintf("ok (moved %s -> %s)", src, dst), nil
		},
	}
}

func makeRenameFileTool(sdPtr *string) Tool {
	return Tool{
		Name:        "rename_file",
		Description: "Rename a file within its directory. Provide the current path and the new filename (basename only, no path separators). For cross-directory moves, use move_file instead.",
		Permission:  PermissionWrite,
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{
					"type":        "string",
					"description": "Current file path",
				},
				"new_name": map[string]any{
					"type":        "string",
					"description": "New filename (basename only, no '/' or '\\' allowed)",
				},
			},
			"required": []string{"path", "new_name"},
		},
		Execute: func(ctx context.Context, args map[string]any) (string, error) {
			raw, _ := args["path"].(string)
			newName, _ := args["new_name"].(string)

			if newName == "" {
				return "", &ToolError{Message: "new_name is required"}
			}
			if strings.ContainsAny(newName, "/\\") {
				return "", &ToolError{Message: "new_name must be a basename only (no path separators)"}
			}

			resolved, err := ResolvePath(raw, *sdPtr)
			if err != nil {
				return "", err
			}

			parent := filepath.Dir(resolved)
			dest := filepath.Join(parent, newName)

			if _, err := os.Stat(resolved); os.IsNotExist(err) {
				return "", &ToolError{Message: fmt.Sprintf("File not found: %s", resolved)}
			}
			if _, err := os.Stat(dest); err == nil {
				return "", &ToolError{Message: fmt.Sprintf("Destination already exists: %s", dest)}
			}

			if err := os.Rename(resolved, dest); err != nil {
				return "", &ToolError{Message: fmt.Sprintf("Failed to rename file: %s", err)}
			}

			return fmt.Sprintf("ok (renamed %s -> %s)", resolved, dest), nil
		},
	}
}

// copyFile copies a file from src to dst (used for cross-device moves).
func copyFile(src, dst string) error {
	s, err := os.Open(src)
	if err != nil {
		return err
	}
	defer s.Close()

	d, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer d.Close()

	_, err = io.Copy(d, s)
	return err
}

File: src/tools/factories/file_io.cpp
#include "tools.h"

#include <git2.h>

#include <filesystem>
#include <fstream>
#include <string>

Tool make_read_file_tool(std::shared_ptr<std::string> safe_dir_ptr,
    const std::vector<std::string>& read_only_paths,
    std::shared_ptr<std::vector<std::string>> tool_logs) {

    Tool t;
    t.name = "read_file";

    t.description = "Read a block of lines from a file."
        " Start with any small range; the discovered total line count will be reported."
        " Lines are prefixed with line numbers, 1-based."
    ;

    t.parameters = {
        {"type", "object"},
        {"properties", {
            {"path", {{"type", "string"}, {"description", "Path to the file to read."}}},
            {"start_line", {{"type", "integer"}, {"description", "Beginning of range, 1-based"}}},
            {"end_line", {{"type", "integer"}, {"description", "End of range (inclusive): 1-based"}}},
        }},
        {"required", {"path","start_line","end_line"}}
    };

    t.execute = [safe_dir_ptr, read_only_paths, tool_logs](const json& args) -> Result<std::string> {

        for (auto& el: args.items()) {
            if (el.key() == "path") continue;
            if (el.key() == "start_line") continue;
            if (el.key() == "end_line") continue;
            return std::unexpected("unknown argument: " + el.key());
        }

        auto raw = args.value("path", std::string());
        auto resolved = resolve_path(raw, *safe_dir_ptr, read_only_paths);
        if (!resolved) {
            return std::unexpected(resolved.error());
        }

        std::error_code ec;
        if (std::filesystem::exists(*resolved, ec) &&
            !std::filesystem::is_regular_file(*resolved, ec)) {
            return std::unexpected("Not a regular file: " + *resolved);
        }

        std::ifstream file(*resolved);
        if (!file.is_open()) {
            return std::unexpected("Failed to open file: " + *resolved);
        }

        std::string result(std::istreambuf_iterator<char>{file}, {});

        int start_line = args.value("start_line",1);
        int end_line = args.value("end_line",1);

        std::stringstream ss;
        ss << "read_file " << raw << '\n' << format_line_range(result, start_line, end_line);
        return ss.str();
    };
    return t;
}

Tool make_write_file_tool(std::shared_ptr<std::string> safe_dir_ptr,
    FileModifiedCallback on_file_modified) {
    Tool t;
    t.name = "write_file";
    t.description = "Write content to a file, creating parent directories if needed";

    t.parameters = {
        {"type", "object"},
        {"properties", {
            {"path", {{"type", "string"}, {"description", "File path"}}},
            {"content", {{"type", "string"}, {"description", "Content to write"}}}
        }},
        {"required", {"path", "content"}}
    };

    t.execute = [safe_dir_ptr, on_file_modified](const json& args) -> Result<std::string> {
        auto raw = args.value("path", std::string());
        auto content = args.value("content", std::string());

        auto resolved = resolve_path(raw, *safe_dir_ptr);
        if (!resolved) {
            return std::unexpected(resolved.error());
        }

        std::error_code ec;
        std::filesystem::create_directories(std::filesystem::path(*resolved).parent_path(), ec);
        if (ec) {
            return std::unexpected("Failed to create parent directories: " + ec.message());
        }

        std::ofstream file(*resolved, std::ios::binary);
        if (!file.is_open()) {
            return std::unexpected("Failed to write file: " + *resolved);
        }
        file.write(content.data(), content.size());
        file.close();

        // Notify the callback that a file was modified
        if (on_file_modified) {
            on_file_modified(*resolved);
        }

        return "ok (" + std::to_string(content.size()) + " bytes written)";
    };
    return t;
}

Tool make_edit_file_tool(std::shared_ptr<std::string> safe_dir_ptr,
    FileModifiedCallback on_file_modified) {
    Tool t;
    t.name = "edit_file";
    t.description = "Edit a file by searching for an exact string and replacing it. "
                    "The search string must match exactly once in the file — this ensures edits "
                    "are safe and unambiguous. "
                    "Use this to make targeted surgical edits instead of rewriting entire files.";

    t.parameters = {
        {"type", "object"},
        {"properties", {
            {"path", {
                {"type", "string"}, {"description", "File path to edit"}}},
            {"search", {
                {"type", "string"},
                {"description",
                    "Exact string to search for; must match exactly once in the file. "
                    "Include surrounding context (unique nearby lines) to guarantee a "
                    "single match."}}},
            {"replace", {
                {"type", "string"},
                {"description", "String to replace the matched occurrence with"}}}
        }},
        {"required", {"path", "search", "replace"}}
    };

    t.execute = [safe_dir_ptr, on_file_modified](const json& args) -> Result<std::string> {
        auto raw = args.value("path", std::string());
        auto search = args.value("search", std::string());
        auto replace = args.value("replace", std::string());

        if (search.empty()) {
            return std::unexpected(std::string("search string is required"));
        }

        auto resolved = resolve_path(raw, *safe_dir_ptr);
        if (!resolved) {
            return std::unexpected(resolved.error());
        }

        std::error_code ec;
        if (std::filesystem::exists(*resolved, ec) &&
            !std::filesystem::is_regular_file(*resolved, ec)) {
            return std::unexpected("Not a regular file: " + *resolved);
        }

        // Read the file
        std::ifstream file(*resolved, std::ios::binary);
        if (!file.is_open()) {
            return std::unexpected("Failed to read file: " + *resolved);
        }
        std::string content(
            (std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        file.close();

        // Count occurrences of the search string
        size_t count = 0;
        size_t pos = 0;
        while ((pos = content.find(search, pos)) != std::string::npos) {
            count++;
            pos += search.size();
        }

        if (count == 0) {
            return std::unexpected("Search string not found in file (0 matches).");
        }
        if (count > 1) {
            return std::unexpected("Search string found " + std::to_string(count) +
                " times in file (expected exactly 1). "
                "Include more surrounding context in the search string to uniquely identify the "
                "location.");
        }

        // Locate the unique occurrence
        pos = content.find(search);

        // Save old content for diff generation before modifying
        std::string old_content = content;

        // Replace the search string with the replacement
        content.replace(pos, search.size(), replace);

        // Write the modified content back to the file
        std::ofstream out(*resolved, std::ios::binary);
        if (!out.is_open()) {
            return std::unexpected("Failed to write file: " + *resolved);
        }
        out.write(content.data(), content.size());
        out.close();

        // Notify the callback that a file was modified
        if (on_file_modified) {
            on_file_modified(*resolved);
        }

        // Compute the line number where the edit occurred (1-indexed)
        int line_num = 1;
        for (size_t i = 0; i < pos; i++) {
            if (content[i] == '\n')
                line_num++;
        }

        // Generate a unified diff of the change using libgit2
        std::string diff_text;
        {
            git_diff_options diff_opts = GIT_DIFF_OPTIONS_INIT;
            auto print_cb = [](const git_diff_delta* /*delta*/,
                               const git_diff_hunk* /*hunk*/,
                               const git_diff_line* line,
                               void* payload) -> int {
                auto* output = static_cast<std::string*>(payload);
                // Prepend origin character for +/-/context lines
                if (line->origin == '+' || line->origin == '-' || line->origin == ' ') {
                    output->push_back(line->origin);
                }
                output->append(line->content, line->content_len);
                return 0;
            };
            int err = git_diff_buffers(
                old_content.data(), old_content.size(), resolved->c_str(),
                content.data(), content.size(), resolved->c_str(),
                &diff_opts,
                nullptr, nullptr, nullptr,
                print_cb, &diff_text);
            if (err) {
                diff_text.clear(); // fall back to no diff
            }
        }

        std::string result = "ok (replaced 1 occurrence at line " + std::to_string(line_num) +
            ", " + std::to_string(search.size()) + " bytes -> " +
            std::to_string(replace.size()) + " bytes)";
        if (!diff_text.empty()) {
            result += " diff follows:\n" + diff_text;
        }
        return result;
    };
    return t;
}


